src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,189 @@
+        private Node<T> prev;
+        private Node<T> next;
+        private T value;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.value = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds!!!");
+        }
+        Node<T> result = first;
+        for (int i = 0; i < index; i++) {
+            result = result.next;
+        }
+        return result.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = node(index);
+        T oldVal = node.value;
+        node.value = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
-    public boolean remove(T object) {
-        return false;
+    public boolean remove(T value) {
+        int index = indexOf(value);
+        if (index < 0) {
+            return false;
+        }
+        remove(index);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index <= size;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException("Index out of bounds!!!");
+        }
+    }
+
+    private void checkElementIndex(int index) {
+        if (!isElementIndex(index)) {
+            throw new IndexOutOfBoundsException("Index out of bounds!!!");
+        }
+    }
+
+    private Node<T> node(int index) {
+        Node<T> node;
+        if (index < (size >> 1)) {
+            node = first;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+        } else {
+            node = last;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+        }
+        return node;
+    }
+
+    private boolean isElementIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private void linkLast(T e) {
+        Node<T> l = last;
```

Comment:
```suggestion
        Node<T> oldLast = last;
```
![image](https://github.com/mate-academy/jv-linked-list/assets/20326858/d7d0a44d-d05a-4f98-9a5a-0bb3e9ccb635)